networkx/algorithms/cycles.py

```diff
@@ -463,3 +466,146 @@ def find_cycle(G, source=None, orientation='original'):
+    necessarily returned in a order by which they appear in the cycle
+
+    Examples
+    --------
+    >>> G=nx.Graph()
+    >>> G.add_cycle([0,1,2,3])
+    >>> G.add_cycle([0,3,4,5])
+    >>> print(nx.minimum_cycle_basis(G))
+    [[0, 1, 2, 3], [0, 3, 4, 5]]
+
+    References:    
+        [1] Kavitha, Telikepalli, et al. "An O(m^2n) Algorithm for Minimum Cycle Basis of Graphs."
+        http://link.springer.com/article/10.1007/s00453-007-9064-z
+        [2] de Pina, J. 1995. Applications of shortest path methods. Ph.D. thesis, University of Amsterdam, Netherlands 
+
+    See Also
+    --------
+    simple_cycles, cycle_basis
+
+    """
+
+    # We first split the graph in commected subgraphs
+    return sum((_min_cycle_basis(c, weight) for c in
+                nx.connected_component_subgraphs(G)), [])
+
+def _min_cycle_basis(comp, weight):
+    cb = []
+    # We  extract the edges not in a spanning tree. We do not really need a 
+    # *minimum* spanning tree. That is why we call the next function with 
+    # weight=None. Depending on implementation, it may be faster as well. 
+    spanning_tree_edges = list(
+        nx.minimum_spanning_edges(comp, weight=None, data=False))
+    edges_excl = [
+        frozenset(e) for e in comp.edges() if e not in spanning_tree_edges]
+    N = len(edges_excl)
+
+    # We maintain a set of vectors orthogonal to sofar found cycles
+    set_orth = [set([edge]) for edge in edges_excl]
+    for k in range(N):
+        # kth cycle is "parallel" to kth vector in set_orth
+        new_cycle = _min_cycle(comp, set_orth[k], weight=weight)
+        cb.append(list(set().union(*new_cycle)))
+        # now update set_orth so that k+1,k+2... th elements are
+        # orthogonal to the newly found cycle, as per [p. 336, 1]
+        base = set_orth[k]
+        set_orth[k + 1:] = [orth ^ base if len(orth & new_cycle) % 2
+                            else orth for orth in set_orth[k + 1:]]
+
+    return cb
+
+
+def _min_cycle(G, orth, weight=None):
+    """
+    Computes the minimum weight cycle in G, orthogonal to the vector orth
+    as per [p. 338, 1]
+    """
+    T = nx.Graph()
+
+    nodes_idx = {node: idx for idx, node in enumerate(G.nodes())}
+    idx_nodes = {idx: node for node, idx in nodes_idx.items()}
+
+    nnodes = len(nodes_idx)
+
+    # Add 2 copies of each edge in G to T. If edge is in orth, add cross edge;
+    # otherwise in-plane edge
+    if weight is not None:
+        for u, v in G.edges():
+            uidx, vidx = nodes_idx[u], nodes_idx[v]
+            edge_w = G[u][v][weight]
+            if frozenset((u, v)) in orth:
+                T.add_edges_from(
+                    [(uidx, nnodes + vidx), (nnodes + uidx, vidx)], weight=edge_w)
+
+            else:
+                T.add_edges_from(
+                    [(uidx, vidx), (nnodes + uidx, nnodes + vidx)], weight=edge_w)
+
+        all_shortest_pathlens = nx.shortest_path_length(T, weight='weight')
```

Comment:
A few more comments sorry. This should be `weight=weight`, not `weight='weight'`.

Also, instead of having essentially duplicate code from line 553 to line 564 and from line 567 to line 577, just do

    for u, v, data in G.edges(data=True):
        edge_w = data.get(weight, 1)
        ... # Put the rest of the body of the (weighted) loop here
    all_shortest_pathlens = nx.shortest_path_length(T, weight=weight)

The result will be the same, because if `weight is None`, then all edges in `T` will end up having weight one.